src/arch/arm/armlib/exceptions/exceptions.S

```diff
@@ -121,14 +122,36 @@ prefetch_abt_handler:
 fiq_handler:
 	sub r14, r14, #4
 	b arm_unresolvable_exception
-data_abt_handler:
 	sub r14, r14, #8
 
 arm_unresolvable_exception:
+	/* This exception may occur if MMU tables or stack pointer
+	 * were corrupted, but we still need to print something, i.e.
+	 * here we try to avoid infinite recoursive exception */
+
+	/* First, make sure MMU is off before accessing stack */
+	mrc p15, 0, r0, c1, c0, 0
+	and r0, r0, #~CR_M
+	mcr p15, 0, r0, c1, c0, 0
+	/* Second, make sure stack pointer has sane value */
+	ldr sp, =_stack_top
 	nop
+	/* Now we are ready to process exception itself */
 	stmfd sp!, {r0-r12, r14}
 	mrs r0, CPSR
 	mrs r1, SPSR
 	stmfd sp!, {r0, r1}
 	mov r0, sp
 	b arm_exception_handler
+
+data_abt_handler:
+	/* Same as with arm_unresolvable_exception we need
+	 * to make sure that RAM access will be safe */
+	mrc p15, 0, r0, c1, c0, 0
+	and r0, r0, #~CR_M
+	mcr p15, 0, r0, c1, c0, 0
+	ldr sp, =_stack_top
+	nop
+	stmfd r13!, {r0-r12,r14}
```

Comment:
Just a small remark. Probably to use the same notation as above? sp instead of r13?